python/Ganga/Core/GangaRepository/Registry.py

```diff
@@ -171,22 +161,45 @@ def __repr__(self):
         return "Incomplete object in '%s', ID %i. Try reload() or remove()." % (self.registry.name, self.id)
 
 
-def synchronised(f):
+def synchronised_flush_lock(f):
+    """
+    Specific flush lock as flushing can take a long time. This ensures no changes occur while flushing though reads can.
+
+    Args:
+        f (func): Function to decorate
+    """
+    @functools.wraps(f)
+    def decorated(self, *args, **kwargs):
+        with self._flush_lock:
+            return f(self, *args, **kwargs)
+    return decorated
+
+def synchronised_read_lock(f):
     """
-    This decorator must be attached to a method on a ``Registry``
-    It uses the object's lock to make sure that the object is held for the duration of the decorated function
+    General read lock for quick functions that won't take a while but won't change anything either
+
     Args:
-        f (function): Function in question being wrapped
+        f (func): Function to decorate
     """
     @functools.wraps(f)
     def decorated(self, *args, **kwargs):
-        # This is the Registry RLock
-        # It it safe in principle to re-enter the registry through other lockable methods in the active lock-holding thread
-        # However, the registry will decide through the use of the transaction locks as to whether it should actually do any work as a result
-        with self._lock:
+        with self._read_lock:
             return f(self, *args, **kwargs)
     return decorated
 
+def synchronised_complete_lock(f):
+    """
+    Entirely lock the registry. Make sure locks are acquired in the right order!
+
+    Args:
+        f (func): Function to decorate
+    """
+    @functools.wraps(f)
+    def decorated(self, *args, **kwargs):
+        with self._flush_lock:
```

Comment:
Not sure what you mean here - this only acquires one lock. If you mean the `@complete_lock' one where both locks are required, the order is determined by what functions can try to acquire what locks, i.e. when flushing, the read lock is required we need to make sure that the complete_lock gets this flush lock **first** to avoid a deadlock.